pallets/mips/src/lib.rs

```diff
@@ -736,78 +785,111 @@ impl<T: Trait> Module<T> {
+    fn close_proposal(index: MipsIndex) {
+        if <Voting<T>>::get(index).is_some() {
+            Self::unreserve_deposits(index);
         }
 
-        if <Proposals<T>>::take(&proposal_hash).is_some() {
-            <Voting<T>>::remove(&proposal_hash);
-            let hash = proposal_hash;
-            <ProposalMetadata<T>>::mutate(|metadata| metadata.retain(|m| m.proposal_hash != hash));
-            <ProposalByIndex<T>>::remove(index);
+        if <Proposals<T>>::take(index).is_some() {
+            <Voting<T>>::remove(index);
+            <ProposalMetadata<T>>::mutate(|metadata| metadata.retain(|m| m.index != index));
 
-            Self::deposit_event(RawEvent::ProposalClosed(index, hash));
+            Self::deposit_event(RawEvent::ProposalClosed(index));
         }
     }
 
-    /// It returns back each deposit to their owners for an specific `proposal_hash`.
-    fn unreserve_deposits(proposal_hash: &T::Hash) {
-        <Deposits<T>>::iter_prefix(proposal_hash).for_each(|depo_info| {
+    /// It returns back each deposit to their owners for an specific `index`.
+    fn unreserve_deposits(index: MipsIndex) {
+        <Deposits<T>>::iter_prefix(index).for_each(|depo_info| {
             let _ = <T as Trait>::Currency::unreserve(&depo_info.owner, depo_info.amount);
         });
-        <Deposits<T>>::remove_prefix(proposal_hash);
+        <Deposits<T>>::remove_prefix(index);
+    }
+
+    fn prepare_to_dispatch(index: MipsIndex) -> DispatchResult {
+        ensure!(
+            <Referendums<T>>::contains_key(index),
+            Error::<T>::MismatchedProposalIndex
+        );
+
+        // Set the default enactment period and move it to `Scheduled`
+        let curr_block_number = <system::Module<T>>::block_number();
+        let enactment_period = curr_block_number + Self::default_enactment_period();
+
+        <Referendums<T>>::mutate(index, |referendum| {
+            if let Some(ref mut referendum) = referendum {
+                referendum.enactment_period = enactment_period;
+                referendum.state = MipsState::Scheduled;
+            }
+        });
+        <ScheduledReferendumsAt<T>>::mutate(enactment_period, |indexes| indexes.push(index));
+
+        Self::deposit_event(RawEvent::ReferendumScheduled(index, enactment_period));
+        Ok(())
     }
 
-    fn prepare_to_dispatch(hash: T::Hash) {
-        if let Some(referendum) = <Referendums<T>>::get(&hash) {
-            let result = match referendum.dispatch(system::RawOrigin::Root.into()) {
-                Ok(_) => true,
+    fn execute_referendum(index: MipsIndex) -> DispatchResult {
+        let referendum = Self::referendums(index).ok_or(Error::<T>::MismatchedProposalIndex)?;
+
+        // Ensure that it is not in its enactment period.
+        let curr_block = <system::Module<T>>::block_number();
+        ensure!(
+            referendum.enactment_period <= curr_block,
+            Error::<T>::ReferendumOnEnactmentPeriod
```

Comment:
Is there ever a case where referendum.enactment_period != curr_block`?

Reply:
You're right. In the latest implementation is not possible.